File: internal/models/monitor_config.go
package models

import "strings"

// Header is a single HTTP header (used by HTTP monitors and by webhooks).
type Header struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

// MonitorConfig holds every type specific option of a monitor. A single flat
// struct is used on purpose: it keeps the JSON column, the API payload and the
// UI form in sync, and Go simply ignores the fields that do not apply to the
// current monitor type.
type MonitorConfig struct {
	// --- HTTP(s) and HTTP(s) Keyword -------------------------------------
	URL                 string   `json:"url,omitempty"`
	Method              string   `json:"method,omitempty"`    // GET POST PUT PATCH DELETE HEAD OPTIONS
	Encoding            string   `json:"encoding,omitempty"`  // json | form | raw | xml
	Body                string   `json:"body,omitempty"`      // request body
	Headers             []Header `json:"headers,omitempty"`   // dynamic key/value list
	AuthType            string   `json:"auth_type,omitempty"` // none | basic | bearer
	BasicUser           string   `json:"basic_user,omitempty"`
	BasicPass           string   `json:"basic_pass,omitempty"`
	BearerToken         string   `json:"bearer_token,omitempty"`
	IgnoreTLS           bool     `json:"ignore_tls,omitempty"`            // ignore TLS/SSL errors
	MaxRedirects        int      `json:"max_redirects,omitempty"`         // 0 = no redirect, default 10
	CacheBuster         bool     `json:"cache_buster,omitempty"`          // append a random up_cachebuster parameter
	AcceptedStatusCodes string   `json:"accepted_status_codes,omitempty"` // "200-299,301"

	// --- Keyword ----------------------------------------------------------
	Keyword       string `json:"keyword,omitempty"`
	InvertKeyword bool   `json:"invert_keyword,omitempty"`
	CaseSensitive bool   `json:"case_sensitive,omitempty"`

	// --- TCP --------------------------------------------------------------
	Host   string `json:"host,omitempty"`
	Port   int    `json:"port,omitempty"`
	Send   string `json:"send,omitempty"`   // optional payload written after connect
	Expect string `json:"expect,omitempty"` // optional expected response (substring)

	// --- DNS --------------------------------------------------------------
	Hostname       string `json:"hostname,omitempty"`        // name to resolve
	ResolverServer string `json:"resolver_server,omitempty"` // 1.1.1.1 or 8.8.8.8:53
	RecordType     string `json:"record_type,omitempty"`     // A AAAA CNAME MX TXT NS SOA
	ExpectedValue  string `json:"expected_value,omitempty"`  // expected value / keyword
	InvertCheck    bool   `json:"invert_check,omitempty"`

	// --- SSL / TLS --------------------------------------------------------
	// ServerName is the SNI sent on the handshake of a ssl monitor (useful when
	// the certificate names a virtual host that the IP alone does not identify).
	ServerName string `json:"server_name,omitempty"`
}

// HTTPMethods lists the methods accepted by the HTTP and Keyword monitors.
var HTTPMethods = []string{"GET", "POST", "PUT", "PATCH", "DELETE", "HEAD", "OPTIONS"}

// HTTPEncodings lists the body encodings accepted by the HTTP monitors.
var HTTPEncodings = []string{"json", "form", "raw", "xml"}

// DNSRecordTypes lists the record types supported by the DNS monitor.
var DNSRecordTypes = []string{"A", "AAAA", "CNAME", "MX", "TXT", "NS", "SOA"}

// Normalize fills in the defaults of the optional fields so the checkers, the
// notifier and the UI can rely on them.
func (c *MonitorConfig) Normalize(monitorType MonitorType) {
	if c.Headers == nil {
		c.Headers = []Header{}
	}
	switch monitorType {
	case MonitorTypeHTTP, MonitorTypeKeyword:
		c.URL = strings.TrimSpace(c.URL)
		c.Method = strings.ToUpper(strings.TrimSpace(c.Method))
		if c.Method == "" {
			c.Method = "GET"
		}
		if c.Encoding == "" {
			c.Encoding = "json"
		}
		if c.AuthType == "" {
			c.AuthType = "none"
		}
		if c.MaxRedirects == 0 {
			c.MaxRedirects = 10
		}
		if strings.TrimSpace(c.AcceptedStatusCodes) == "" {
			c.AcceptedStatusCodes = "200-299"
		}
	case MonitorTypeDNS:
		c.Hostname = strings.TrimSpace(c.Hostname)
		c.RecordType = strings.ToUpper(strings.TrimSpace(c.RecordType))
		if c.RecordType == "" {
			c.RecordType = "A"
		}
		if strings.TrimSpace(c.ResolverServer) == "" {
			c.ResolverServer = "1.1.1.1"
		}
	case MonitorTypeTCP:
		c.Host = strings.TrimSpace(c.Host)
	case MonitorTypeSSL:
		c.Host = strings.TrimSpace(c.Host)
		c.ServerName = strings.TrimSpace(c.ServerName)
		if c.Port == 0 {
			c.Port = DefaultSSLPort
		}
	}
}

// PruneToType returns a copy of the configuration keeping only the fields the
// given type reads.
//
// It is the Go twin of `CONFIG_FIELDS` in `web/src/lib/monitor-config.ts` and is
// what lets a bulk row override the type of its template: without it the options
// of the template type would travel into a monitor of another type (the checker
// ignores them, but they would show up in the stored JSON and in the API payload
// of a monitor they do not describe).
func (c MonitorConfig) PruneToType(monitorType MonitorType) MonitorConfig {
	switch monitorType {
	case MonitorTypeHTTP:
		return MonitorConfig{
			URL: c.URL, Method: c.Method, Encoding: c.Encoding, Body: c.Body,
			Headers: c.Headers, AuthType: c.AuthType, BasicUser: c.BasicUser,
			BasicPass: c.BasicPass, BearerToken: c.BearerToken, IgnoreTLS: c.IgnoreTLS,
			MaxRedirects: c.MaxRedirects, CacheBuster: c.CacheBuster,
			AcceptedStatusCodes: c.AcceptedStatusCodes,
		}
	case MonitorTypeKeyword:
		return MonitorConfig{
			URL: c.URL, Method: c.Method, Encoding: c.Encoding, Body: c.Body,
			Headers: c.Headers, AuthType: c.AuthType, BasicUser: c.BasicUser,
			BasicPass: c.BasicPass, BearerToken: c.BearerToken, IgnoreTLS: c.IgnoreTLS,
			MaxRedirects: c.MaxRedirects, CacheBuster: c.CacheBuster,
			AcceptedStatusCodes: c.AcceptedStatusCodes,
			Keyword:             c.Keyword, InvertKeyword: c.InvertKeyword, CaseSensitive: c.CaseSensitive,
		}
	case MonitorTypeTCP:
		return MonitorConfig{Host: c.Host, Port: c.Port, Send: c.Send, Expect: c.Expect}
	case MonitorTypeDNS:
		return MonitorConfig{
			Hostname: c.Hostname, ResolverServer: c.ResolverServer,
			RecordType: c.RecordType, ExpectedValue: c.ExpectedValue, InvertCheck: c.InvertCheck,
		}
	case MonitorTypeSSL:
		return MonitorConfig{Host: c.Host, Port: c.Port, ServerName: c.ServerName, IgnoreTLS: c.IgnoreTLS}
	}
	return MonitorConfig{}
}

// DefaultSSLPort is the port a ssl monitor dials when none is configured.
const DefaultSSLPort = 443

// HeaderMap converts the dynamic header list into a map, ignoring empty keys.
func (c *MonitorConfig) HeaderMap() map[string]string {
	out := make(map[string]string, len(c.Headers))
	for _, h := range c.Headers {
		key := strings.TrimSpace(h.Key)
		if key == "" {
			continue
		}
		out[key] = h.Value
	}
	return out
}

File: web/src/lib/monitor-config.ts
import type { MonitorConfig, MonitorPayload, MonitorType } from './types'

/**
 * Type aware pruning of a monitor payload.
 *
 * The API speaks one flat config struct plus a handful of top level switches
 * (the certificate and domain watches, the template link) and the form renders
 * the same object for every probe type: only the fields of the selected type
 * are shown, but the hidden ones keep the values of the type the monitor had.
 * Sending them back is what the backend rejects ("cert_watch is only available
 * for http, keyword and ssl monitors"), so every save goes through
 * `sanitizeMonitorPayload`, which drops what does not belong to the type the
 * operator selected.
 *
 * `CONFIG_FIELDS` mirrors the Go side (`internal/models/monitor_config.go`, the
 * `Validate` of every type in `internal/services/monitor_validate.go` and the
 * `Supports*` helpers in `internal/models/enums.go`): keep the two in sync when
 * a monitor type or one of its options is added (see `docs/architecture.md`).
 *
 * Fields shared by several types (`host`/`port` for TCP and SSL, `ignore_tls`
 * for HTTP, Keyword and SSL) are listed by each of them on purpose, so a switch
 * between those types does not lose the value the new one still uses.
 */
export const CONFIG_FIELDS: Record<MonitorType, readonly (keyof MonitorConfig)[]> = {
  http: [
    'url',
    'method',
    'encoding',
    'body',
    'headers',
    'auth_type',
    'basic_user',
    'basic_pass',
    'bearer_token',
    'ignore_tls',
    'max_redirects',
    'cache_buster',
    'accepted_status_codes',
  ],
  keyword: [
    'url',
    'method',
    'encoding',
    'body',
    'headers',
    'auth_type',
    'basic_user',
    'basic_pass',
    'bearer_token',
    'ignore_tls',
    'max_redirects',
    'cache_buster',
    'accepted_status_codes',
    'keyword',
    'invert_keyword',
    'case_sensitive',
  ],
  tcp: ['host', 'port', 'send', 'expect'],
  dns: ['hostname', 'resolver_server', 'record_type', 'expected_value', 'invert_check'],
  ssl: ['host', 'port', 'server_name', 'ignore_tls'],
}

/** supportsCertificate reports whether a type can read a TLS certificate. */
export function supportsCertificate(type: MonitorType): boolean {
  return type === 'http' || type === 'keyword' || type === 'ssl'
}

/**
 * supportsDomainWatch reports whether a type has a target a registrable domain
 * can be derived from. Every current type does; the switch exists because the
 * backend has the same one (`MonitorType.SupportsDomainWatch`).
 */
export function supportsDomainWatch(type: MonitorType): boolean {
  switch (type) {
    case 'http':
    case 'keyword':
    case 'tcp':
    case 'dns':
    case 'ssl':
      return true
  }
  return false
}

/** pruneMonitorConfig keeps only the probe options the type actually reads. */
export function pruneMonitorConfig(type: MonitorType, config?: MonitorConfig | null): MonitorConfig {
  const source = config ?? {}
  const pruned: Record<string, unknown> = {}
  for (const field of CONFIG_FIELDS[type]) {
    const value = source[field]
    if (value === undefined || value === null) continue
    pruned[field] = value
  }
  return pruned as MonitorConfig
}

/** What the sanitizer needs to know about the template the monitor follows. */
export interface SanitizeOptions {
  /**
   * Type of the linked template, when the caller has the template list. A link
   * that points at a template of another type is not a template this monitor
   * can follow, so it is dropped instead of failing the save.
   */
  templateType?: MonitorType | null
}

/**
 * sanitizeMonitorPayload returns a copy of the payload without the fields that
 * do not apply to the selected type: the probe options, the certificate
 * switches and the template link. The form can keep those values hidden while
 * the operator experiments with the type; they never reach the API.
 */
export function sanitizeMonitorPayload(payload: MonitorPayload, options: SanitizeOptions = {}): MonitorPayload {
  const clean: MonitorPayload = { ...payload }
  const type = (clean.type ?? 'http') as MonitorType

  clean.config = pruneMonitorConfig(type, clean.config)

  // The certificate switches only exist for the types that can read a
  // certificate, and notifying about a watch that is off is a silent no-op.
  if (!supportsCertificate(type) || !clean.cert_watch) {
    clean.cert_watch = false
    clean.cert_notify = false
    clean.cert_warn_days = ''
  }

  // The domain watch is available everywhere, but nothing can outlive it.
  if (!clean.domain_watch) {
    clean.domain_notify = false
    clean.domain_warn_days = ''
    clean.domain_expires_at = null
  }

  if (clean.template_uuid && options.templateType && options.templateType !== type) {
    clean.template_uuid = ''
  }

  return clean
}
